reports/tasks/msisdn_message_report.py

```diff
@@ -13,63 +16,200 @@ class GenerateMSISDNMessageReport(BaseTask):
+                continue
+
+            data[msisdn]['reg_date'] = registration.created_at.strftime(
+                "%Y-%m-%d %H:%M:%S")
+            data[msisdn]['msg_type'] = registration.data.get('msg_type', "")
+            data[msisdn]['preg_week'] = registration.data.get('preg_week', "")
+
+            # Get facility info from the operator's identity
+            operator_id = registration.data.get('operator_id', None)
+            if operator_id is not None:
+                operator_identity = is_client.get_identity(operator_id)
+                data[msisdn]['facility'] = operator_identity.get(
+                        'details', {}).get('facility_name', "")
+            else:
+                logger.info(
+                    'No operator_id on registration for {0}'.format(msisdn))
+                data[msisdn]['facility'] = ""
+
+        return data
+
+    def retrieve_messages(self, ms_client, data, start_date, end_date):
+        logger = self.get_logger()
+
+        longest_list = 0
+        for msisdn in data:
+            message_list = []
+
+            if data[msisdn].get('id', None) is None:
+                # Skip if we didn't find an identity
+                continue
+
+            response = ms_client.get_outbounds({
+                "to_identity": data[msisdn]['id'],
+                "after": start_date.strftime("%Y-%m-%dT00:00:00"),
+                "before": end_date.strftime("%Y-%m-%dT00:00:00")
             })
+            results = list(response['results'])
+
+            if len(results) < 1:
+                logger.info(
+                    'No results from message sender for {0}'.format(msisdn))
+
+                data[msisdn]["messages"] = message_list
+                continue
+
+            for message in results:
+                message_list.append({
+                    "content": message['content'],
+                    # Reformat the date
+                    "date_sent": message['created_at'].replace('T', ' ')[:-8],
+                    "status": 'Delivered' if message['delivered'] else 'Undelivered'  # noqa
+                })
+
+            data[msisdn]["messages"] = message_list
+            if len(message_list) > longest_list:
+                longest_list = len(message_list)
+
+        return (data, longest_list)
+
+    def populate_spreadsheet(self, msisdns, data, list_length):
+        workbook = ExportWorkbook()
+        sheet = workbook.add_sheet('Data for study cohort', 0)
+
+        header = [
+            'Phone number',
+            'Date registered',
+            'Facility',
+            'Pregnancy week',
+            'Message type'
+        ]
+
+        i = 1
+        while i <= list_length:
+            header.extend([
+                'Message %d: content' % i,
+                'Message %d: date sent' % i,
+                'Message %d: status' % i
+            ])
+            i += 1
+
+        sheet.set_header(header)
+
+        for msisdn in msisdns:
```

Comment:
For any msisdn that doesn't have a related identity, we've set the value in `data` to an empty dictionary. So we can probably iterate over `data.items()` here, and we'll still get the full list of msisdns.

Reply:
I'm not sure if the ordering is important or not, but it's not a big thing, so happy to keep it this way to maintain the ordering.